app/course/controllers/CourseCtrl.js

```diff
@@ -23,6 +23,135 @@ courseApp.controller('CourseCtrl', ['$scope', '$rootScope', '$routeParams', '$ti
 				academicYear: String(i).yearToAcademicYear()
 			});
 		}
+
+		$scope.toggleAssignTagsDropdown = function() {
+			if ($scope.view.isAssignTagsDropdownOpen) {
+				$scope.closeAssignTagsDropdown();
+			} else {
+				$scope.view.isAssignTagsDropdownOpen = true;
+			}
+		};
+
+		$scope.closeAssignTagsDropdown = function() {
+			$scope.view.isAssignTagsDropdownOpen = false;
+			$scope.clearTagUserChoices();
+		};
+
+		$scope.clearTagUserChoices = function() {
+			$scope.view.state.tags.availableIds.forEach(function(tagId) {
+				$scope.view.tagOccurences[tagId].userChoice = "none";
+				$scope.view.tagOccurences[tagId].icon = $scope.view.tagOccurences[tagId].presence;
+			});
+		};
+
+		$scope.calculateTagStates = function() {
```

Comment:
Comment this method. What's a tag state? What concerns exist around the way they work?